flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java

```diff
@@ -0,0 +1,301 @@
+		int numberOfLocalChannelsGate1 = 0;
+		int numberOfRemoteChannelsGate2 = 1;
+		int numberOfLocalChannelsGate2 = 1;
+
+		int numberOfRemoteChannelsTotal = numberOfRemoteChannelsGate1 + numberOfRemoteChannelsGate2;
+		int numberOfInputGates = 2;
+
+		int buffersPerChannel = 2;
+		int extraNetworkBuffersPerGate = 8;
+
+		CloseableRegistry closeableRegistry = new CloseableRegistry();
+
+		NettyShuffleEnvironment network = new NettyShuffleEnvironmentBuilder()
+			.setNetworkBuffersPerChannel(buffersPerChannel)
+			.setFloatingNetworkBuffersPerGate(extraNetworkBuffersPerGate)
+			.build();
+
+		Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple1 = buildInputGate(
+			network,
+			numberOfRemoteChannelsGate1,
+			numberOfLocalChannelsGate1);
+		Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple2 = buildInputGate(
+			network,
+			numberOfRemoteChannelsGate2,
+			numberOfLocalChannelsGate2);
+
+		SingleInputGate inputGate1 = tuple1.f0;
+		SingleInputGate inputGate2 = tuple2.f0;
+
+		try {
+			assertEquals(tuple1.f1.size(), numberOfRemoteChannelsGate1 + numberOfLocalChannelsGate1);
+			RemoteInputChannel remoteInputChannel1 = tuple1.f1.get(0);
+			RemoteInputChannel remoteInputChannel2 = tuple1.f1.get(1);
+
+			SingleInputGate[] inputGates = new SingleInputGate[]{tuple1.f0, tuple2.f0};
+			FloatingBuffersUsageGauge floatingBuffersUsageGauge = new FloatingBuffersUsageGauge(inputGates);
+			ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
+			CreditBasedInputBuffersUsageGauge inputBuffersUsageGauge = new CreditBasedInputBuffersUsageGauge(
+				floatingBuffersUsageGauge,
+				exclusiveBuffersUsageGauge,
+				inputGates);
+
+			assertEquals(0.0, exclusiveBuffersUsageGauge.getValue(), 0.0);
+			assertEquals(0.0, inputBuffersUsageGauge.getValue(), 0.0);
+
+			drainBuffer(buffersPerChannel, remoteInputChannel1, closeableRegistry);
+
+			int totalBuffers = extraNetworkBuffersPerGate * numberOfInputGates + buffersPerChannel * numberOfRemoteChannelsTotal;
+
+			assertEquals(buffersPerChannel, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
+			assertEquals((double) buffersPerChannel / (buffersPerChannel * numberOfRemoteChannelsTotal), exclusiveBuffersUsageGauge.getValue(), 0.0001);
+			assertEquals((double) buffersPerChannel / totalBuffers, inputBuffersUsageGauge.getValue(), 0.0001);
+
+			drainBuffer(buffersPerChannel, remoteInputChannel2, closeableRegistry);
+
+			assertEquals(buffersPerChannel * numberOfRemoteChannelsGate1, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
+			assertEquals((double) buffersPerChannel * numberOfRemoteChannelsGate1 / (buffersPerChannel * numberOfRemoteChannelsTotal),
+				exclusiveBuffersUsageGauge.getValue(), 0.0001);
+			assertEquals((double) buffersPerChannel * numberOfRemoteChannelsGate1 / totalBuffers,
+				inputBuffersUsageGauge.getValue(), 0.0001);
+		} finally {
+			inputGate1.close();
```

Comment:
The relevant closes could also be added into `CloseableRegistry` as the first test done.

Reply:
As i mentioned above to @pnowojski , I think the closeableRegistry do not respect to the register order which will lead to some buffer have not be released with some excpetion

Reply:
And the closeableRegistry will close quietly, ignoring the exception (exclusive buffers may have not been released before close global buffer pool). So I prefer to keep the current shape.